src/osd/osd_types.h

```diff
@@ -3959,6 +3996,34 @@ static inline ostream& operator<<(ostream& out, const notify_info_t& n) {
 	     << " " << n.timeout << "s)";
 }
 
+struct object_manifest_t {
+  enum {
+    TYPE_NONE = 0,
+    TYPE_REDIRECT = 1,  // start with this
+    TYPE_CHUNKED = 2,   // do this later
+  };
+  uint8_t type;  // redirect, chunked, ...
+  ghobject_t redirect_target;
+
+  object_manifest_t() : type(0) { }
+  object_manifest_t(uint8_t type, const ghobject_t& redirect_target) 
+    : type(type), redirect_target(redirect_target) { }
+
+  bool is_empty() const {
+    return type == TYPE_NONE;
+  }
+  bool is_redirect() const {
+    return type == TYPE_REDIRECT;
+  }
+  bool is_chunked() const {
+    return type == TYPE_CHUNKED;
+  }
+  static void generate_test_instances(list<object_manifest_t*>& o);
+  void encode(bufferlist &bl) const;
+  void decode(bufferlist::iterator &bl);
+  void dump(Formatter *f) const;
+};
+WRITE_CLASS_ENCODER(object_manifest_t)
```

Comment:
For this patch, please include just the object_manifest_t and object_info_t changes and leave off the extensible_mode, maybe_handle_*, mon, and other changes.

Also, for object_manifest_t, please add a static const char *get_type_name(int t) and const char *get_type_name() accessor that returns "none", "chunked", etc., and also add an friend ostream& operator<<(...).  And in the object_info_t operator<< add "if (has_manifest()) out << " " << manifest;"